src/Pim/Component/Catalog/Validator/Constraints/UniqueValueValidator.php

```diff
@@ -42,38 +41,28 @@ public function __construct(ProductRepositoryInterface $repository, UniqueValues
      * It means that we make this validator stateful which is a bad news, the good one is we ensure this validation
      * for any processes (other option was to mess the import as we did with previous implementation)
      *
-     * Due to constraint guesser, the constraint is applied on :
-     * - ProductValueInterface data when applied through form
-     * - ProductValueInterface when applied directly through validator
+     * Due to constraint guesser, the constraint is applied on ProductValueInterface when applied
+     * directly through validator.
      *
      * The constraint guesser should be re-worked in a future version to avoid such behavior
      *
-     * @param ProductValueInterface|mixed $data
-     * @param Constraint                  $constraint
+     * @param ProductValueInterface $productValue
+     * @param Constraint            $constraint
      *
      * @see Pim\Bundle\CatalogBundle\Validator\ConstraintGuesser\UniqueValueGuesser
      */
-    public function validate($data, Constraint $constraint)
+    public function validate($productValue, Constraint $constraint)
     {
-        // TODO: This validator will be completely reworked in TIP-698
-        return;
-
-        if (empty($data)) {
+        if (empty($productValue)) {
             return;
         }
 
-        if (is_object($data) && $data instanceof ProductValueInterface) {
-            $productValue = $data;
-        } else {
-            $productValue = $this->getProductValueFromForm();
-        }
-
         if ($productValue instanceof ProductValueInterface && $productValue->getAttribute()->isUnique()) {
-            $valueAlreadyExists = $this->alreadyExists($productValue);
+            $valueAlreadyExists = $this->alreadyExists($productValue, $this->context->getRoot());
```

Comment:
we check that this value does not exist in another product
because the value can already be set on this particular product if we didn't change it